### PR TITLE
Improve Flatten to avoid using dynamic shapes when possible

### LINKED ISSUE #25402: Flatten is not constant folded properly by constfold grappler.
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 16.04
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): master (1/31/19)
- Python version: 3.5.2
- Bazel version (if compiling from source): 0.19.2
- GCC/Compiler version (if compiling from source): c++ (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
- CUDA/cuDNN version: CUDA 10.0.130
- GPU model and memory: Quadro RTX 6000, 24GiB

**Describe the current behavior**
tf.layers.Flatten uses Shape, StridedSlice, Pack ops to build the new shape for Reshape. These ops are not constant folded.

**Describe the expected behavior**
 Since the shape of `x` is known at compile time, the constant folder should be able to fold away these ops to create a single Const for the new shape for Reshape.

**Code to reproduce the issue**
```python
import tensorflow as tf
import math
from tensorflow.core.protobuf import meta_graph_pb2
from tensorflow.core.protobuf import rewriter_config_pb2
from tensorflow.python.grappler import tf_optimizer

def build_graph(x, output_name='output'):
  # x.shape = (None, 1, 1, 1536)
  x = tf.layers.Flatten()(x)
  # This should perform exactly the same as:
  # x = tf.reshape(x, [-1, 1536])
  return tf.identity(x, name=output_name)

def apply_constfold(frozen_graph, output_nodes):
  graph = tf.Graph()
  with graph.as_default():
    tf.import_graph_def(frozen_graph, name="")
  grappler_meta_graph_def = tf.train.export_meta_graph(graph_def=graph.as_graph_def(add_shapes=True), graph=graph)

  _to_bytes = lambda s: s.encode("utf-8", errors="surrogateescape")
  output_collection = meta_graph_pb2.CollectionDef()
  output_list = output_collection.node_list.value
  for i in output_nodes:
    if isinstance(i, tf.Tensor):
      output_list.append(_to_bytes(i.name))
    else:
      output_list.append(_to_bytes(i))
  # TODO(laigd): use another key as the outputs are really not train_op.
  grappler_meta_graph_def.collection_def["train_op"].CopyFrom(output_collection)
  rewriter_config = rewriter_config_pb2.RewriterConfig()
  rewriter_config.optimizers.extend(["constfold"])

  session_config_with_trt = tf.ConfigProto()
  session_config_with_trt.graph_options.rewrite_options.CopyFrom(
      rewriter_config)
  frozen_graph = tf_optimizer.OptimizeGraph(session_config_with_trt, grappler_meta_graph_def, graph_id=b"tf_graph")
  return frozen_graph

if __name__ == '__main__':
  with tf.Graph().as_default():
    # Create graph
    x = tf.placeholder(dtype=tf.float32, shape=(None, 1, 1, 1536), name='input')
    y = build_graph(x)
    # Initialize
    with tf.Session() as sess:
      sess.run(tf.global_variables_initializer())
      # Freeze graph
      frozen_graph = tf.graph_util.convert_variables_to_constants(
          sess,
          sess.graph_def,
          output_node_names=['output'])

  print('Nodes before:')
  [print(n.name, n.op) for n in frozen_graph.node]

  # const folding
  frozen_graph = apply_constfold(frozen_graph, output_nodes=['output'])

  print('----------------------------------------')
  print('Nodes after:')
  [print(n.name, n.op) for n in frozen_graph.node]
```

**Output**
```
Nodes before:
input Placeholder
flatten/Shape Shape
flatten/strided_slice/stack Const
flatten/strided_slice/stack_1 Const
flatten/strided_slice/stack_2 Const
flatten/strided_slice StridedSlice
flatten/Reshape/shape/1 Const
flatten/Reshape/shape Pack
flatten/Reshape Reshape
output Identity
-------------------------
Nodes after:
input Placeholder
flatten/strided_slice/stack Const
flatten/strided_slice/stack_1 Const
flatten/strided_slice/stack_2 Const
flatten/Reshape/shape/1 Const
flatten/Shape Shape
flatten/strided_slice StridedSlice
flatten/Reshape/shape Pack
flatten/Reshape Reshape
output Identity
```

** Expected Output **
```
Nodes before:
input Placeholder
flatten/Shape Shape
flatten/strided_slice/stack Const
flatten/strided_slice/stack_1 Const
flatten/strided_slice/stack_2 Const
flatten/strided_slice StridedSlice
flatten/Reshape/shape/1 Const
flatten/Reshape/shape Pack
flatten/Reshape Reshape
output Identity
-------------------------
Nodes after:
input Placeholder
Reshape/shape Const
Reshape Reshape
output Identity
```

### PR DESCRIPTION
**Pull request by trevor-m**
Flatten currently creates a reshape which introduces a dependency on the size of the batch dimension. In the cases where flatten is commonly used, the batch dimension is usually not known statically but the rest of the dimensions are. 
This means that the constant folding grappler pass cannot resolve the shape ahead of time. This also makes it difficult to convert using TF-TRT.

The implementation of Flatten can be rewritten such that there is no dependency on an unknown batch dimension.

Previous implementation (simplification):
```
def flatten(x):
  return tf.reshape(x, [tf.shape(x)[0], -1])
```
This implementation:
```
def flatten(x):
  if all(x.shape[1:]):
    return tf.reshape(x, shape=(-1, prod(x.shape[1:]))
  else:
    return tf.reshape(x, shape=(tf.shape(x)[0], -1))
```

Fixes #25402

### CODE REVIEW

**Comment by rthadur** [comment]
@ebrevdo gentle ping to review PR..!

**Comment by tfboyd** [comment]
@ebrevdo   Can you take a look or let us know if we should reassign?

File: tensorflow/python/keras/layers/core.py [hunk 1]
```diff
@@ -580,9 +580,13 @@ def call(self, inputs):
       permutation.append(1)
       inputs = array_ops.transpose(inputs, perm=permutation)
 
-    outputs = array_ops.reshape(
-        inputs, (tensor_shape.dimension_value(inputs.shape[0]) or
-                 array_ops.shape(inputs)[0], -1))
+    input_shape = tensor_shape.TensorShape(inputs.shape).as_list()
```
Comment (ebrevdo) [inline tensorflow/python/keras/layers/core.py]: replace with:

```python
input_shape = inputs.shape
if inputs_shape[1:].is_fully_defined():
  outputs = array_ops.reshape(inputs, (-1,tensor_shape.dimension_value(np.prod(inputs_shape[1:]]))
else:
  ...
```

**Comment by tanzhenyu** [comment]
I added a label here just in case no one from keras is assigned to this.

**Comment by ebrevdo** [comment]
Actually; I'd first check that shape[0].is_fully_defined() because that's a
faster check than shape[1:].is_fully_defined().  swap the if/else
components.

**Comment by trevor-m** [comment]
> Actually; I'd first check that shape[0].is_fully_defined() because that's a faster check than shape[1:].is_fully_defined(). swap the if/else components.

But shape[0] being undefined doesn't mean than shape[1:] will be fully defined. We can only do this optimization when shape[1:] is fully defined.

**Comment by aaroey** [comment]
@ebrevdo would you please help to take a look at this PR? Thanks.

**Comment by ebrevdo** [comment]
I see failures like [this one](https://source.cloud.google.com/results/invocations/1895d420-b447-4d78-b679-0f421227eb03/targets/%2F%2Ftensorflow%2Fpython:layers_core_test/tests).  Probably what's happening is you are passing an np.prod() on an empty array is an empty float32 array.  you need to cast the result to int64, say, for that special case.

**Comment by trevor-m** [comment]
> I see failures like [this one](https://source.cloud.google.com/results/invocations/1895d420-b447-4d78-b679-0f421227eb03/targets/%2F%2Ftensorflow%2Fpython:layers_core_test/tests). Probably what's happening is you are passing an np.prod() on an empty array is an empty float32 array. you need to cast the result to int64, say, for that special case.

Thanks @ebrevdo, I had fixed those tests in [this commit](https://github.com/tensorflow/tensorflow/pull/30380/commits/2f61f75e244891e9ce1d10fa3a34fd4cb419a5d4) by passing new dim size with `int()`. Is this an acceptable solution?

**Review by ebrevdo — APPROVED** [review]

**Comment by ebrevdo** [comment]
We had to roll this PR back, likely due to an integer overflow in the reshape op:

```python
inputs = tf.zeros(shape=[8, 21316, 21316, 80])
tf.reshape(inputs,[-1, 36349748480])
`
gives an error:
"Dimension size must be evenly divisible by 1990010112 but is 290797987840 for 'Reshape_2' (op: 'Reshape') with input shapes: [8,21316,21316,80], [2] and with input tensors computed as partial shapes: input[1] = [?,1990010112]."

Note that 1990010112 = np.int32(36349748480).
```

Looks like somewhere in ReshapeV2 we're not handling the int64 case properly.

Probably we can

**Comment by ebrevdo** [comment]
Actually the bug looks like it's somewhere in the shape inference code.  Trying to track it down.

**Comment by ebrevdo** [comment]
Can you resubmit with int64 casting?  e.g., replace the shape tuple with a casted version:

```
math_ops.cast((-1, flattened_dim), dtypes.int64)
```

**Comment by ebrevdo** [comment]
(only cast if flattened_dim is > np.iinfo(np.int32).max)

**Comment by ebrevdo** [comment]
[this would be a temporary solution, we're working on an internal bugfix since the bug seems deeper than just your use case]

also instead of math_ops.cast, use constant_op.constant()

**Comment by trevor-m** [comment]
Thanks @ebrevdo for investigating that bug! Here's a new PR with your requested changes: https://github.com/tensorflow/tensorflow/pull/31450
